### ISSUE TITLE
Document tf.Dataset.zip/map triggers tf.Dataset.shuffle

### ISSUE THREAD

**Issue by Aweptimum** [tensorflow/tensorflow#70521]
I just got bit by this and was wondering if it's intended or needs a warning label

@aaudiber in #41334 commented the below:

`tf.data.Dataset` objects don't eagerly compute all of their data. They work like blueprints, where the data is computed on the fly every time you iterate through the dataset. As a result, iterating through the same dataset multiple times could result in different output.

The `Dataset.zip` transformation takes multiple datasets and ***iterates*** through them in parallel. If you want to iterate through the input just once, use `map` instead of `zip`:
...

_Originally posted by @aaudiber in https://github.com/tensorflow/tensorflow/issues/41334#issuecomment-662747527_

I had a workflow like so for a tf.data pipeline:
```python
filenames, labels = getFileLabels()
loaded = loadFiles(filenames)
transformed = transformData(loaded)
labeled = tf.data.Dataset.zip(transformed, labels).shuffle()
```
I was just relying on the `filenames`, `labels`, and `transformed` datasets being in the same order. To avoid loading all data into RAM, I moved shuffling up to just the filenames

```python
filenames, labels = getFileLabels()
filenames = filenames.shuffle()
loaded = loadFiles(filenames)
transformed = transformData(loaded)
labeled = tf.data.Dataset.zip(transformed, labels)
```
And what I found was that the model's accuracy tanked because it was receiving data with random labels. It's not explicitly stated anywhere in the docs, but it seems calls to `map` and `zip`(and anything else that would iterate) trigger `shuffle`.

The workaround is easy - just map the loading over the shuffled filenames and return a tuple directly rather than relying on `zip` - but originally I was just going off someone's advice to shuffle "pointers" to files rather than the actual loaded files (which made a lot of sense). Because I don't actually know what's happening at a low-level in the execution graph, I had just assumed that `shuffle` wouldn't be triggered by its sibling methods, only by `model.fit`

**TL;DR**:  
I think it should be documented how `shuffle` can impact other operations, particularly `zip`

**Comment by tilakrayal** [comment]
@Aweptimum,
Can you please share a reproducible code that supports your statement so that the issue can be debug more effectively and also if possible could you try to contribute with the PR for the changes to be made in the document. Thank you!

**Comment by Aweptimum** [comment]
I haven't had a chance to make a MRE, but I will when I can

**Comment by Aweptimum** [comment]
Here's a MRE:
```python
import tensorflow as tf

filenames = [
    ['foo'],
    ['bar'],
    ['baz'],
]

data = [
    [1],
    [2],
    [3]
]

## Expected
fileset = tf.data.Dataset.from_tensor_slices(filenames)
dataset = tf.data.Dataset.from_tensor_slices(data)

zipped = tf.data.Dataset.zip(fileset, dataset).shuffle(3)
lst = list(zipped.as_numpy_iterator())
print("\nExpected")
print(lst)

## Actual
fileset = tf.data.Dataset.from_tensor_slices(filenames).shuffle(3)
dataset = tf.data.Dataset.from_tensor_slices(data)

zipped = tf.data.Dataset.zip(fileset, dataset)
lst = list(zipped.as_numpy_iterator())
print("\nActual")
print(lst)
```

Output
![image](https://github.com/tensorflow/tensorflow/assets/40273116/d5453a6d-452d-4f45-b6dd-a675a90587b3)

Friendlier output:
Shuffled together:  ('foo', 1), ('bar', 2), ('baz', 3)
Shuffled filenames: ('foo', 3), ('bar', 1), ('baz', 2)

After writing this MRE I can see now that this is probably just an idiosyncrasy of `shuffle`, which executes eagerly from what I can tell (if I'm wrong let me know). I had originally thought that `shuffle` would not execute until the training loop, when iteration started, whereas `zip` immediately changed the structure of the dataset. This made it appear to me that `zip` was triggering `shuffle` to run (hence the issue title), when in reality shuffle is running as soon as it's called.

TL;DR - There's no bug, it's just that what the documentation states and what I observed didn't add up to the real answer.

I think a distinction between lazy and eager dataset methods should be drawn - whether they are evaluated immediately or in the training loop (I had assumed everything was lazy). `shuffle`'s documentation doesn't indicate that is is executed immediately. I was going off of `reshuffle_each_iteration`, which implies `shuffle` is executed in epoch's (the training loop), and so I thought the "buffer" it referred to was allocated once iteration started.

**Comment by Venkat6871** [comment]
Hi @Aweptimum,
Thanks for the detailed explanation and for sharing the reproducible example that really helped clarify the behavior.
As you noted, this is expected behavior: each dataset in `tf.data.Dataset.zip` is iterated independently, so shuffling one dataset before zipping can break the correspondence between elements. To reduce confusion, a documentation update has been proposed to clarify how `shuffle` interacts with `zip`:
https://github.com/tensorflow/tensorflow/pull/111928
Thanks to @VibhorGautam for contributing the fix  
We will keep this issue open until the PR is reviewed and merged.

Thanks again for raising this it will help improve the documentation.
